### PR TITLE
fix(acls): auth bypass via forward auth

### PR DESCRIPTION
**Pull request by scottmckendry**
## Summary by CodeRabbit

- **Bug Fixes**
  - Improved handling of forwarded request paths for access control decisions.
  - Allow/block rules now correctly match intended prefixes and avoid substring/query-string mismatches.
  - Consistent root-path behavior for allow and block scenarios.
  - Invalid forwarded URIs and malformed path rules are safely rejected (deny by default).
- **Tests**
  - Expanded coverage for forwarded-path allow/block edge cases, including query-string handling and encoded/substring variations.

### CODE REVIEW

File: internal/service/access_controls_rules.go [hunk 1]
```diff
@@ -180,33 +181,65 @@ type AuthEnabledRule struct {
...
+			continue
+		}
+
+		if strings.HasPrefix(path, configuredPath) {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
```
Comment (steveiliop56) [inline internal/service/access_controls_rules.go:212]: @scottmckendry we are changing behavior here and possibly breaking things. The way ACLs for paths should work is the following:

## Normal whitelist

In case we have something like:

```
tinyauth.apps.foo.path.allow: /bar,/foo/bar,/hello
```

On every request we loop through the rules and check if the request URI has any prefix that matches. For example, `/bar` and all sub-paths such as `/bar/foo/bar` will be allowed. This is the current **intended** behavior.

## Regex whitelist

For regex whitelist there is no looping, the check for regex happens before the comma split and we just check if the rule starts and ends with slashes (`/`). For example this example is allowed:

```
tinyauth.apps.foo.path.allow: /\/foo|bar/
```

This is **not** allowed:

```
tinyauth.apps.foo.path.allow /\/foo|bar/,/\/test/
```

There is no need to allow such thing since your comma-seperated whitelist can be recreated using regex and users using regex will know what they are doing. Now, as for how regex is treated, it's a simple match (not a prefix match). It's the user's responsibility to know how to configure regex matching correctly, not ours.

Hopefully this clears things out a bit on how the parsing of the whitelist should work.

File: internal/controller/proxy_controller.go [hunk 2]
```diff
@@ -390,6 +390,16 @@ func (controller *ProxyController) getForwardAuthContext(c *gin.Context) (ProxyC
...
 
+	parsedURI, err := url.ParseRequestURI(uri)
+
+	if err != nil {
+		return ProxyContext{}, fmt.Errorf("invalid x-forwarded-uri: %w", err)
+	}
+
+	if parsedURI.Path == "" {
+		parsedURI.Path = "/"
+	}
```
Comment (steveiliop56) [inline internal/controller/proxy_controller.go]: We should do this for all auth module types and also strip out the query parameters immediately. Tinyauth won't need them in any part of the authentication chain anyway.

**Review by steveiliop56 — CHANGES_REQUESTED** [review]

File: internal/service/access_controls_rules.go (hunk 1 above)
Comment (scottmckendry) [reply internal/service/access_controls_rules.go:212]: thanks, this is quite helpful actually. I think I've covered all scenarios now.

File: internal/controller/proxy_controller.go (hunk 2 above)
Comment (scottmckendry) [reply internal/controller/proxy_controller.go]: good call, done :)

File: internal/service/access_controls_rules.go [hunk 3]
```diff
@@ -180,33 +181,56 @@ type AuthEnabledRule struct {
...
+		}
+
+		return regex.MatchString(path), nil
+	}
+
+	for _, configuredPath := range strings.Split(paths, ",") {
+		if strings.HasPrefix(path, strings.TrimSpace(configuredPath)) {
+			return true, nil
+		}
+	}
```
[the remark this answers is not in the document]
Comment (steveiliop56) [reply internal/service/access_controls_rules.go:209]: @scottmckendry although this is 100% user error let's just apply the suggestion, doesn't cost much code-wiese.

[the remark this answers is not in the document]
Comment (scottmckendry) [reply internal/service/access_controls_rules.go:209]: fixed in latest commit

**Comment by steveiliop56** [comment]
Closing in favor of #1055, thanks @scottmckendry.
